### ISSUE TITLE
Permit Environment Template

### ISSUE THREAD

**Issue by gemanor** [permitio/permit-cli#20]
We want to allow users to simply ready-made policy templates in their environment. Examples of templates are GitHub, E-commerce, Banking, etc.

Permit has a terraform provider that allows developers to apply a whole policy configuration on an environment.
The idea is to have a `templates` folder in this repository containing various templates of common policy configuration (blog, CMS, e-commerce, banking, etc.). By running this command, we will apply the plan to the environment.

To eliminate the need to install Terraform locally, the command will call a lightweight server that will get the current API key as an authorization header and the plan in the body of the request to run the TF commands. To support that, the feature should include a Docker file and a lightweight server that will be used to apply the plan. The server will have a maximum of two files (Dockerfile and server.py/server.js/nginx).sh, etc.) and will sit in a terraform_server folder at the root of this project. We will deploy this server so the CLI will just call an API to apply it.

The command should have two subcommands:
- `permit env template apply` - will print the list of available templates and apply by choice. Command arguments:
    -  `--key` - Optional: use a key instead of the one provided by the `AuthProvider`
    - `--local` - Optional: run the TF commands locally instead of calling the server (will fail if TF isn't installed)
    - `--template`  - Optional: skip the template choice step and apply one of the templates. It will fail if the template does not exist.
- `permit env template list` - will list all the available templates in the template folder. 
    -  `--key` - Optional: use a key instead of the one provided by the `AuthProvider`

Acceptance criteria:
- Commands, Server, and Templates are implemented and tested
- 90% UT coverage on all the CLI code

**Comment by Harsh9485** [comment]
What kind of template do you need?

**Comment by gemanor** [comment]
The idea is to use the Terraform Permit Provider, create some template, and let the user choose one and apply it on environment.
https://github.com/permitio/terraform-provider-permit-io

**Comment by Abiji-2020** [comment]
@gemanor could you provide more details on it .

**Comment by Abiji-2020** [comment]
@gemanor i am interested to work on this feature, let me initially check on setup of the lightweight server and proceed further

**Comment by Abiji-2020** [comment]
@gemanor currently are there any templates to proceed in terraform or we need to create them. ?

**Comment by gemanor** [comment]
There are some:
https://github.com/permitio/mesa-verde-banking-demo/blob/main/main.tf
https://github.com/permitio/fga-tradeoffs/blob/main/main.tf
https://github.com/permitio/terraform-provider-permit-io/blob/main/examples/rebac/main.tf

Let me know if there's any misisng information here.

Also, if I'm not avaialbe here, please reach out to me in Permit community. You can join there via: io.permit.io/slack

**Comment by Abiji-2020** [comment]
Currently I am testing with the terraform provider and checking out how can we do it..... And let me try to create  a template for this examples. And proceed further..

**Comment by gemanor** [comment]
Thank you.

The server util that should do the TF command should be very simple and fail in any error. The idea here is that the command will just follow the "happy path" and not try to deal with conflicts, etc. It should be about 10 lines of code or so.

Please let me know if you have any specific questions here.

**Comment by Abiji-2020** [comment]
> Thank you.
> 
> The server util that should do the TF command should be very simple and fail in any error. The idea here is that the command will just follow the "happy path" and not try to deal with conflicts, etc. It should be about 10 lines of code or so.
> 
> Please let me know if you have any specific questions here.

I thought that we need a standard server and will have more complete and it could handle error so i was testing through like on the design...

**Comment by Abiji-2020** [comment]
The server just will act as a remote which executes the file config we are passing

**Comment by gemanor** [comment]
Exactly 

I don't want to enforce people to install Terraform 

The acceptance criteria of this server for the task, is that we have a docker file that you can run with docker run... and it will do the job for yiy

**Comment by Abiji-2020** [comment]
https://github.com/Abiji-2020/Permit-terraform-test


so this is the server repo I have started working on it... 

In this currently only the actions are added.. will update on the next mile stone @gemanor

**Comment by gemanor** [comment]
Hey, looks like a good start. Not sure what is the first half of the TF template is doing.. the template will sent in the request body.

**Comment by Abiji-2020** [comment]
> Hey, looks like a good start. Not sure what is the first half of the TF template is doing.. the template will sent in the request body.

Ah currently this haves only half the example of the REBAC example .... And let me add more details so that you can get more clear understanding on this.

**Comment by Abiji-2020** [comment]
@gemanor I have changed my approach that we are sending the template TF files in the payload and change the Modify the APIKEY value at the server and apply them.

**Comment by gemanor** [comment]
Hey, it looks very good to me 😊 

We just need to think how we manage it with the main package.json, so it will not compile it..

**Comment by Abiji-2020** [comment]
> Hey, it looks very good to me 😊
> 
> We just need to think how we manage it with the main package.json, so it will not compile it..

I tried building the complete code it doesn't compile it.. It only compiles the files inside the `source` directory. 

![Image](https://github.com/user-attachments/assets/96080232-e1e0-4f62-b8bc-1262492abd21)

**Comment by gemanor** [comment]
Sounds good!

Do you want to proceed with the command itself?

**Comment by Abiji-2020** [comment]
> Sounds good!
> 
> Do you want to proceed with the command itself?

Yes, I am creating another branch for the command and we can store the `.tf` files in the template folder and proceed further right? And in addition to that do we need to have customisable templates?

**Comment by gemanor** [comment]
What do you mean by customisable templates?
